script to check NuttX release candidates (SHA512, GPG sigs, required files)

adamfeuer opened the pull request:
## Summary

* Adds a bash shell  `check-nuttx-release.sh` script that checks a release candidate SHA512, GPG signatures, and required files.
* Automates some manual steps in [Validating a NuttX release](https://cwiki.apache.org/confluence/display/NUTTX/Validating+a+staged+Release)
* Can take a release name or URL
* Leaves the tar.gz files in current directory for user to untar and build

## Impact

* None if you don't run the script

## Testing

* Tested on the 9.1.0-RC1 (`https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/`)
* You can test the script as follows:
```bash
mkdir tmp
cd tmp/
check-nuttx-release.sh https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/
```
## Usage
```bash
./check-nuttx-release.sh -h
Usage: ./check-nuttx-release.sh <URL-of-release-candidate-directory-or-release-name>
   Given release name or release full URL, downloads all files in
   in that directory (which should include nuttx and nuttx-apps
   sha512, asc, and tar.gz files, checks their SHA512 and GPG 
   signatures, and checks the unpacked directories for required 
   files. Creates a temporary directory to do its work in.

   nuttx and nuttx-apps tar.gz files are left in the current
   directory.
```

## Limitations and TODO

* script output could be much cleaner
* does not currently run Apache RAT

btashton commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,101 @@
+
+URL=$1
+
+function download_release() {
```
You can probably use the wget recursive function to do this for you for example
```
wget -r -np -R "index.html*" https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/
```

btashton replied:
Sorry I missed the cut-dir and no host flags
```
wget -r -np -nH --cut-dirs=5 -R "index.html*" https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/
❯ tree
.
├── 9.1.0-RC1
│   ├── apache-nuttx-9.1.0-incubating.tar.gz
│   ├── apache-nuttx-9.1.0-incubating.tar.gz.asc
│   ├── apache-nuttx-9.1.0-incubating.tar.gz.sha512
│   ├── apache-nuttx-apps-9.1.0-incubating.tar.gz
│   ├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc
│   └── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512
└── robots.txt
```

btashton commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,101 @@
+
+URL=$1
+
+function download_release() {
```
You can probably use the wget recursive function to do this for you for example
```
wget -r -np -nH --cut-dirs=5 -R "index.html*" https://dist.apache.org/repos/dist/dev/incubator/nuttx/9.1.0-RC1/
❯ tree
.
├── 9.1.0-RC1
│   ├── apache-nuttx-9.1.0-incubating.tar.gz
│   ├── apache-nuttx-9.1.0-incubating.tar.gz.asc
│   ├── apache-nuttx-9.1.0-incubating.tar.gz.sha512
│   ├── apache-nuttx-apps-9.1.0-incubating.tar.gz
│   ├── apache-nuttx-apps-9.1.0-incubating.tar.gz.asc
│   └── apache-nuttx-apps-9.1.0-incubating.tar.gz.sha512
└── robots.txt
```

adamfeuer replied:
@btashton Good idea, thanks. I updated the script. I had to also use `--cut-dirs` to avoid a deeply nested directory hierarchy.

adamfeuer replied:
@btashton If you're happy with this, I'll rebase and force push, let me know. Or else let me know if you want to see other improvements.

btashton replied:
@adamfeuer Yeah sorry I forgot that and updated it later.  There is also the `-nH` flag I added, but it looks like you are using the host as the temporary directory, so that works as well.  Any reason to keep the full URL as the input instead of just the release name?  Anyway I'll mark this as approved and leave it to your discretion.  We should probably also consider adding some traps or a lazy `set -e` to capture the failure.

adamfeuer replied:
@btashton I just used the URL since that's what you posted in your email :slightly_smiling_face:  I'm open to change it, let me know if you want that.

Re: traps, yeah, I think so to. This just started as a quick and dirty way to avoid manual work... so how about we do the traps in another PR?

Re: lazy `set -e`, what do you mean? I think I don't understand.

btashton replied:
if you use `set -e` then if any of the commands have a non 0 status code the script will fail.  Not all commands return 0 on success, so it is a way to lazy handle it if they do.  The URL is fine, we can make this better in a future release, so lets start here :)

adamfeuer replied:
@btashton Ah, I just didn't understand what you meant by "lazy". I'll add `set -e` and force push. Yes, let's get it in and make it better later.

Ouss4 replied:
I like the idea of passing only the release name, the rest of the URL doesn't change.

adamfeuer replied:
@Ouss4 Ok, I added the ability to do either URL or release name.
@btashton I added the trap statement too.

Ok, will force push.

Ouss4 commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,95 @@
+#!/bin/bash
```
License header missing.

adamfeuer replied:
@Ouss4 will fix

Ouss4 commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,96 @@
+function check_nuttx() {
+    # check nuttx sha512 and gpg
+    RELEASE_FILE="$(ls *.tar.gz|head -1)"
+    check_sha512 "$RELEASE_FILE" check_gpg "$RELEASE_FILE" 
+    check_gpg "$RELEASE_FILE"
```
```suggestion
    check_sha512 "$RELEASE_FILE"
    check_gpg "$RELEASE_FILE"
```
Was that a copy/paste typo?

adamfeuer replied:
@Ouss4 yes, fixed.

adamfeuer commented:
@btashton @Ouss4 Thanks for the improvements! Here's what I did:

* added the `set -e`
* added the trap to remove the tempdir
* added a license header
* removed the cut and paste error

I force pushed, this should be good to go. Let's do more improvements in another PR. :)

Ouss4 commented:
@adamfeuer It doesn't look like you pushed all these new changes.

adamfeuer commented:
@Ouss4 Weird, everything looks up to date on my local git. Looking in to it.

adamfeuer commented:
@Ouss4 Not sure what was wrong. I rebased my changes onto master and force-pushed them here. This should be good to go.

Thanks for catching that, and I'm sorry about the confusion.

Ouss4 commented:
The license needs to be ALv2.  Here is the ASF header: https://github.com/apache/incubator-nuttx/blob/993591dca14a822b48be44513fa01d7ad10e0c61/sched/sched/sched_getcpu.c#L2-L19
Could you change that?

adamfeuer commented:
@Ouss4 Thanks for catching the license header thing– I'll fix it tomorrow, I don't have time tonight. I'll also add APL 2.0 headers to my checklist of things to do before I submit a PR! 🙂

adamfeuer commented:
@Ouss4 I changed license header to APL 2.0. Is it ready to merge now?

xiaoxiang781216 commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,129 @@
+trap "rm -rf $TEMPDIR" EXIT
+
+function download_release() {
+    rm -rf "$TEMPDIR"
```
Don't need?

adamfeuer replied:
@xiaoxiang781216 Yes, you're right, this was here before @btashton suggested using the trap. I will remove it.

xiaoxiang781216 commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,129 @@
+
+function check_nuttx_apps() {
+    # check nuttx-apps sha512 and gpg
+    RELEASE_FILE="$(ls *.tar.gz|head -2| tail -1)"
```
Don't need head -2?

adamfeuer replied:
@xiaoxiang781216 I think this is needed- I'm trying to get the second line of the ls output, not the last line.

xiaoxiang781216 commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,129 @@
+#!/bin/bash
+#############################################################################
+# tools/check-nuttx-release.sh
```
How about change the script name to checkrelease.sh like checkpatch.sh?

Ouss4 replied:
I actually thought about this too.  I'm glad someone else had the same idea.

adamfeuer replied:
@xiaoxiang781216 @Ouss4 I will change the script name

adamfeuer commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,129 @@
+    echo "   files. Creates a temporary directory to do its work in."
+    echo
+    echo "   nuttx and nuttx-apps tar.gz files are left in the current"
+    echo "   directory."
```
@xiaoxiang781216 I will add examples

adamfeuer commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,129 @@
+
+download_release
+check_nuttx 
+check_nuttx_apps 
```
@xiaoxiang781216 This was an attempt at a quick script to reduce some manual steps— I propose leaving this out of this PR, and we can add building the sim:nsh in another PR. How does that sound?

adamfeuer commented on tools/check-nuttx-release.sh:
```diff
@@ -0,0 +1,129 @@
+download_release
+check_nuttx 
+check_nuttx_apps 
+cd "$ORIGINAL_DIR"
```
@xiaoxiang781216 will remove this

adamfeuer commented:
@xiaoxiang781216 I'm proposing building `nsh:sim` in a future PR to keep this one simple. Is that ok with you?

@btashton @Ouss4 @xiaoxiang781216 Ok, I pushed the suggested changes. Does this look good? If so, would one of you be willing to merge?

xiaoxiang781216 commented:
> @xiaoxiang781216 I'm proposing building `nsh:sim` in a future PR to keep this one simple. Is that ok with you?
> 
> @btashton @Ouss4 @xiaoxiang781216 Ok, I pushed the suggested changes. Does this look good? If so, would one of you be willing to merge?

Sure, it is fine now.
